core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -1282,12 +1282,19 @@ object Parser {
   /** product with the first argument being a Parser
     */
   def product10[A, B](first: Parser[A], second: Parser0[B]): Parser[(A, B)] =
-    Impl.Prod(first, second)
+    first match {
+      case f @ Impl.Fail() => f.widen
+      case f @ Impl.FailWith(_) => f.widen
```

Comment:
There are a few more cases below

Reply:
Curious why you like this change? I prefer the opposite. I think the compile to the same code don't they?

I don't like general reflective matching and to a reviewer the  `:` style might be that, so they have to read a bit carefully.

But I'm interested to know your view on it.

Reply:
You got me curious as well because my main motivation for this was based on the fact that this generated different bytecode before. However, after a quick check with this use case, I remembered things badly and they do in fact generate the same bytecode. That said, I'm totally ok with it 😄 

You might be right on it needing careful reading but the current one imo has a slight "disadvantage" of keeping track of the number of args. It's not worth changing because of this.